core/shared/src/main/scala/zio/ZRef.scala

```diff
@@ -15,182 +15,107 @@
 
-  /**
-   * Writes a new value to the `Ref` without providing a guarantee of
-   * immediate consistency.
-   *
-   * @param a value to be written to the `Ref`
-   * @return `IO[EA, Unit]`
-   */
   def setAsync(a: A): IO[EA, Unit]
 
-  /**
-   * Atomically modifies the `Ref` with the specified function. This is not
-   * implemented in terms of `modify` purely for performance reasons.
-   *
-   * @param f function to atomically modify the `Ref`
-   * @return `IO[E, Unit]`
-   */
-  def update[E](f: B => A)(implicit ev1: EA <:< E, ev2: EB <:< E): IO[E, Unit] =
-    modify(b => ((), f(b)))
-
-  /**
-   * Atomically modifies the `Ref` with the specified function and returns the
-   * updated value. This is not implemented in terms of `modify` purely for
-   * performance reasons.
-   *
-   * @param f function to atomically modify the `Ref`
-   * @return `IO[E, B]` modified value of the `Ref`
-   */
-  def updateAndGet[E](f: B => A)(implicit ev1: EA <:< E, ev2: EB <:< E): IO[E, B] =
-    modify(b => (b, f(b)))
-
-  /**
-   * Atomically modifies the `Ref` with the specified partial function.
-   * If the function is undefined on the current value it doesn't change it.
-   *
-   * @param pf partial function to atomically modify the `Ref`
-   * @return `IO[E, Unit]`
-   */
-  def updateSome[E](pf: PartialFunction[B, A])(implicit ev1: EA <:< E, ev2: EB <:< E): IO[E, Unit]
-
-  /**
-   * Atomically modifies the `Ref` with the specified partial function.
-   * If the function is undefined on the current value it returns the old value
-   * without changing it.
-   *
-   * @param pf partial function to atomically modify the `Ref`
-   * @return `IO[E, B]` modified value of the `Ref`
-   */
-  def updateSomeAndGet[E](pf: PartialFunction[B, A])(implicit ev1: EA <:< E, ev2: EB <:< E): IO[E, B]
+  final def collect[C](pf: PartialFunction[B, C]): ZRef[EA, Option[EB], A, C] =
+    fold(identity, Some(_), Right(_), pf.lift(_).fold[Either[Option[EB], C]](Left(None))(Right(_)))
+
+  final def contramap[C](f: C => A): ZRef[EA, EB, C, B] =
+    contramapEither(c => Right(f(c)))
+
+  final def contramapEither[EC >: EA, C](f: C => Either[EC, A]): ZRef[EC, EB, C, B] =
+    dimapEither(f, Right(_))
+
+  final def dimap[C, D](f: C => A, g: B => D): ZRef[EA, EB, C, D] =
+    dimapEither(c => Right(f(c)), b => Right(g(b)))
+
+  final def dimapEither[EC >: EA, ED >: EB, C, D](f: C => Either[EC, A], g: B => Either[ED, D]): ZRef[EC, ED, C, D] =
+    fold(identity, identity, f, g)
+
+  final def dimapError[EC, ED](f: EA => EC, g: EB => ED): ZRef[EC, ED, A, B] =
+    fold(f, g, Right(_), Right(_))
+
+  final def map[C](f: B => C): ZRef[EA, EB, A, C] =
+    mapEither(b => Right(f(b)))
 
+  final def mapEither[EC >: EB, C](f: B => Either[EC, C]): ZRef[EA, EC, A, C] =
+    dimapEither(Right(_), f)
+
+  final def unifyError[E](ea: EA => E, eb: EB => E): ZRef[E, E, A, B] =
+    dimapError(ea, eb)
+
+  final def unifyValue[C](ca: C => A, bc: B => C): ZRef[EA, EB, C, C] =
```

Comment:
All these look like a good start. I'd also add:

```scala
readOnly
writeOnly
```

So it's easy to hand off a "read only" or "write only" view downstream.

Reply:
Done. Let me know if there are any others you can think of. It is very easy to add them.